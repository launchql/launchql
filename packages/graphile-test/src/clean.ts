const uuidRegexp = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

const idReplacement = (v: unknown): string | unknown => (!v ? v : '[ID]');

// Generic object type for any key-value mapping
type AnyObject = Record<string, any>;

function mapValues<T extends AnyObject, R = any>(
  obj: T,
  fn: (value: T[keyof T], key: keyof T) => R
): Record<keyof T, R> {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    acc[key as keyof T] = fn(value, key as keyof T);
    return acc;
  }, {} as Record<keyof T, R>);
}

export const pruneDates = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) => {
    if (!v) {
      return v;
    }
    if (v instanceof Date) {
      return '[DATE]';
    } else if (
      typeof v === 'string' &&
      /(_at|At)$/.test(k as string) &&
      /^20[0-9]{2}-[0-9]{2}-[0-9]{2}/.test(v)
    ) {
      return '[DATE]';
    }
    return v;
  });

export const pruneIds = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) =>
    (k === 'id' || (typeof k === 'string' && k.endsWith('_id'))) &&
    (typeof v === 'string' || typeof v === 'number')
      ? idReplacement(v)
      : v
  );

export const pruneIdArrays = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) =>
    typeof k === 'string' && k.endsWith('_ids') && Array.isArray(v)
      ? `[UUIDs-${v.length}]`
      : v
  );

export const pruneUUIDs = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) => {
    if (typeof v !== 'string') {
      return v;
    }
    if (['uuid', 'queue_name'].includes(k as string) && uuidRegexp.test(v)) {
      return '[UUID]';
    }
    if (k === 'gravatar' && /^[0-9a-f]{32}$/i.test(v)) {
      return '[gUUID]';
    }
    return v;
  });

export const pruneHashes = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) =>
    typeof k === 'string' &&
    k.endsWith('_hash') &&
    typeof v === 'string' &&
    v.startsWith('$')
      ? '[hash]'
      : v
  );

export const prune = (obj: AnyObject): AnyObject =>
  pruneHashes(pruneUUIDs(pruneIds(pruneIdArrays(pruneDates(obj)))));

export const snapshot = (obj: unknown): unknown => {
  if (Array.isArray(obj)) {
    return obj.map(snapshot);
  } else if (obj && typeof obj === 'object') {
    return mapValues(prune(obj as AnyObject), snapshot);
  }
  return obj;
};
