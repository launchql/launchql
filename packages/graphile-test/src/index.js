export * from './graphql';
export * from './gql';
export * from './env';
export * from './clean';
