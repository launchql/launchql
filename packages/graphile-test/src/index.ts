export * from './clean';
export * from './graphile-test';
export * from './connect';
export * from './types';
export { seed } from 'pgsql-test';