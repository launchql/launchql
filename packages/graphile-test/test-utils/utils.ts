export const logDbSessionInfo = async (db: { query: (sql: string) => Promise<any> }) => {
    const res = await db.query(`
      select
        current_user,
        session_user,
        current_setting('role', true) as role,
        current_setting('myapp.user_id', true) as user_id
    `);
    console.log('[db session info]', res.rows[0]);
  };
  