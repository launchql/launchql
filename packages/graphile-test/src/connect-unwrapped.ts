import { getConnections as getPgConnections } from 'pgsql-test';
import { GraphQLTest } from './graphile-test';

import type { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import type { PgTestClient } from 'pgsql-test/test-client';
import type { SeedAdapter } from 'pgsql-test/seed/types';
import type {
  GraphQLQueryOptions,
  GraphQLQueryUnwrappedFn,
  GraphQLQueryUnwrappedFnPos,
  GraphQLResponse
} from './types';

export interface GetConnectionsInput {
  useRoot?: boolean;
  schemas: string[];
  authRole?: string;
}

const unwrap = <T>(res: GraphQLResponse<T>): T => {
  if (res.errors?.length) {
    throw new Error(JSON.stringify(res.errors, null, 2));
  }
  if (!res.data) {
    throw new Error('No data returned from GraphQL query');
  }
  return res.data;
};

export const getConnectionsUnwrapped = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryUnwrappedFn;
  queryPositional: GraphQLQueryUnwrappedFnPos;
}> => {
  const conn: GetConnectionResult = await getPgConnections(input, seedAdapters);
  const { pg, db, teardown: dbTeardown } = conn;

  const gqlContext = GraphQLTest(input, conn);
  await gqlContext.setup();

  const teardown = async () => {
    await gqlContext.teardown();
    await dbTeardown();
  };

  const query: GraphQLQueryUnwrappedFn = async (opts) => unwrap(await gqlContext.query(opts));

  const queryPositional: GraphQLQueryUnwrappedFnPos = async (query, variables, commit, reqOptions) =>
    unwrap(await gqlContext.query({ query, variables, commit, reqOptions }));

  return {
    pg,
    db,
    teardown,
    query,
    queryPositional
  };
};
