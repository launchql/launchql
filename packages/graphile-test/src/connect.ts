import { getConnections as getPgConnections } from 'pgsql-test';
import { GraphQLTest } from './graphile-test';
import type { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import type { PgTestClient } from 'pgsql-test/test-client';
import type { SeedAdapter } from 'pgsql-test/seed/types';
import type { DocumentNode, ExecutionResult } from 'graphql';

export type GraphQLQueryFn<T = ExecutionResult> = (
  query: string | DocumentNode,
  variables?: Record<string, any>,
  commit?: boolean,
  reqOptions?: Record<string, any>
) => Promise<T>;
export interface GetConnectionsInput {
  useRoot?: boolean;
  schemas: string[];
  authRole?: string;
}

/**
 * Combines PostgreSQL test setup with GraphQL test context
 */
export const getConnections = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  
  const conn: GetConnectionResult = await getPgConnections(input, seedAdapters);
  const { pg, db, teardown: dbTeardown } = conn;

  const gqlContext = GraphQLTest(input, conn);
  
  await gqlContext.setup();

  const teardown = async () => {
    await gqlContext.teardown();
    await dbTeardown();
  };

  const query: GraphQLQueryFn = (query, variables, commit, reqOptions) =>
    gqlContext.query({ query, variables, commit, reqOptions });

  return {
    pg,
    db,
    teardown,
    query
  };
};
