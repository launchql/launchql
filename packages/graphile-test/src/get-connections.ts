import { getConnections as getPgConnections } from 'pgsql-test';
import { GraphQLTest } from './graphile-test';

import type { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import type { PgTestClient } from 'pgsql-test/test-client';
import type { SeedAdapter } from 'pgsql-test/seed/types';
import type {
  GraphQLQueryOptions,
  GraphQLQueryFnObj,
  GraphQLQueryFn,
  GraphQLQueryUnwrappedFnObj,
  GraphQLQueryUnwrappedFn,
  GraphQLResponse,
  GetConnectionsInput
} from './types';

// Core unwrapping utility
const unwrap = <T>(res: GraphQLResponse<T>): T => {
  if (res.errors?.length) {
    throw new Error(JSON.stringify(res.errors, null, 2));
  }
  if (!res.data) {
    throw new Error('No data returned from GraphQL query');
  }
  return res.data;
};

// Base connection setup - shared across all variants
const createConnectionsBase = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
) => {
  const conn: GetConnectionResult = await getPgConnections(input, seedAdapters);
  const { pg, db, teardown: dbTeardown } = conn;

  const gqlContext = GraphQLTest(input, conn);
  await gqlContext.setup();

  const teardown = async () => {
    await gqlContext.teardown();
    await dbTeardown();
  };

  const baseQuery = (opts: GraphQLQueryOptions) => gqlContext.query(opts);
  const baseQueryPositional = (query: any, variables?: any, commit?: boolean, reqOptions?: any) =>
    gqlContext.query({ query, variables, commit, reqOptions });

  return {
    pg,
    db,
    teardown,
    baseQuery,
    baseQueryPositional
  };
};

// ============================================================================
// REGULAR QUERY VERSIONS
// ============================================================================

/**
 * Creates connections with raw GraphQL responses
 */
export const getConnectionsObject = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  return {
    pg,
    db,
    teardown,
    query: baseQuery
  };
};

/**
 * Creates connections with unwrapped GraphQL responses (throws on errors)
 */
export const getConnectionsObjectUnwrapped = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryUnwrappedFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryUnwrappedFnObj = async (opts) => unwrap(await baseQuery(opts));

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with logging for GraphQL queries
 */
export const getConnectionsObjectWithLogging = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFnObj = async (opts) => {
    console.log('Executing GraphQL query:', opts.query);
    const result = await baseQuery(opts);
    console.log('GraphQL result:', result);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with timing for GraphQL queries
 */
export const getConnectionsObjectWithTiming = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFnObj = async (opts) => {
    const start = Date.now();
    const result = await baseQuery(opts);
    const duration = Date.now() - start;
    console.log(`GraphQL query took ${duration}ms`);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};

// ============================================================================
// POSITIONAL QUERY VERSIONS
// ============================================================================

/**
 * Creates connections with raw GraphQL responses (positional API)
 */
export const getConnections = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  return {
    pg,
    db,
    teardown,
    query: baseQueryPositional
  };
};

/**
 * Creates connections with unwrapped GraphQL responses (positional API, throws on errors)
 */
export const getConnectionsUnwrapped = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryUnwrappedFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryUnwrappedFn = async (query, variables, commit, reqOptions) =>
    unwrap(await baseQueryPositional(query, variables, commit, reqOptions));

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with logging for GraphQL queries (positional API)
 */
export const getConnectionsWithLogging = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFn = async (query, variables, commit, reqOptions) => {
    console.log('Executing positional GraphQL query:', query);
    const result = await baseQueryPositional(query, variables, commit, reqOptions);
    console.log('GraphQL result:', result);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with timing for GraphQL queries (positional API)
 */
export const getConnectionsWithTiming = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFn = async (query, variables, commit, reqOptions) => {
    const start = Date.now();
    const result = await baseQueryPositional(query, variables, commit, reqOptions);
    const duration = Date.now() - start;
    console.log(`Positional GraphQL query took ${duration}ms`);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};