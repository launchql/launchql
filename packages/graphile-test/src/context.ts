import { graphql, print, ExecutionResult, DocumentNode } from 'graphql';
import { withPostGraphileContext, PostGraphileOptions } from 'postgraphile';
// @ts-ignore
import MockReq from 'mock-req';
import type { Client, Pool } from 'pg';
import { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import { GetConnectionsInput } from './connect';

interface PgSettings {
    [key: string]: string;
}

export const runGraphQLInContext = async <T = ExecutionResult>({
    input,
    conn,
    pgPool,
    schema,
    options,
    authRole,
    query,
    variables,
    reqOptions = {}
}: {
    input: GetConnectionsInput & GetConnectionOpts,
    conn: GetConnectionResult;
    pgPool: Pool;
    schema: any;
    options: PostGraphileOptions;
    authRole: string;
    query: string | DocumentNode;
    variables?: Record<string, any>;
    reqOptions?: Record<string, any>;
}): Promise<T> => {
    if (!conn.pg.client) {
        throw new Error('pgClient is required and must be provided externally.');
    }

    const req = new MockReq({
        url: options.graphqlRoute || '/graphql',
        method: 'POST',
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json'
        },
        ...reqOptions
    });

    const pgSettingsGenerator = options.pgSettings;
    // @ts-ignore
    const pgSettings: PgSettings =
        typeof pgSettingsGenerator === 'function'
            ? await pgSettingsGenerator(req)
            : pgSettingsGenerator || {};

    // @ts-ignore
    return await withPostGraphileContext(   
        { ...options, pgPool, pgSettings },
        async context => {

            const pgConn = input.useRoot ? conn.pg : conn.db;
            const pgClient = pgConn.client;
            await setContextOnClient(pgClient, pgSettings, authRole);
            await pgConn.ctxQuery();

            const printed = typeof query === 'string' ? query : print(query);
            const result = await graphql({
                schema,
                source: printed,
                contextValue: { ...context, pgClient },
                variableValues: variables ?? null
            });
            return result as T;
        }
    );
};


export async function setContextOnClient(
  pgClient: Client,
  pgSettings: Record<string, string>,
  role: string
): Promise<void> {
  await pgClient.query(`select set_config('role', $1, true)`, [role]);

  for (const [key, value] of Object.entries(pgSettings)) {
    await pgClient.query(`select set_config($1, $2, true)`, [key, String(value)]);
  }
}
