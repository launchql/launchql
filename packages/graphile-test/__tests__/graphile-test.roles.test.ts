process.env.LOG_SCOPE = 'graphile-test';

import { getConnectionsPositional as getConnections } from '../src/get-connections';
import { snapshot } from '../src';
import { seed } from 'pgsql-test';
import { join } from 'path';
import gql from 'graphql-tag';
import type { GraphQLQueryFn, GraphQLQueryFnPos } from '../src/types';
import type { PgTestClient } from 'pgsql-test/test-client';
import { logDbSessionInfo } from '../test-utils/utils';

const schemas = ['app_public'];
const sql = (f: string) => join(__dirname, '/../sql', f);

let teardown: () => Promise<void>;
let query: GraphQLQueryFnPos;
let db: PgTestClient;

beforeAll(async () => {
  const connections = await getConnections(
    {
      schemas,
      authRole: 'authenticated'
    },
    [
      seed.sqlfile([
        sql('test.sql'),
        sql('grants.sql')
      ])
    ]
  );

  ({ db, query, teardown } = connections);
});

beforeEach(() => db.beforeEach());

beforeEach(async () => {
  db.setContext({
    role: 'authenticated',
    'myapp.user_id': '123'
  });
});

afterEach(() => db.afterEach());

afterAll(async () => {
  await teardown();
});

// ✅ Basic mutation and query test
it('creates a user and fetches it', async () => {
  await logDbSessionInfo(db);
  const CREATE_USER = gql`
    mutation CreateUser($input: CreateUserInput!) {
      createUser(input: $input) {
        user {
          id
          username
        }
      }
    }
  `;

  const GET_USERS = gql`
    query {
      users {
        nodes {
          id
          username
        }
      }
    }
  `;

  const newUsername = 'testuser1';

  const createRes = await query(CREATE_USER, {
    input: {
      user: {
        username: newUsername
      }
    }
  });

  expect(snapshot(createRes)).toMatchSnapshot('createUser');

  const fetchRes: any = await query(GET_USERS);

  expect(
    fetchRes.data.users.nodes.some((u: any) => u.username === newUsername)
  ).toBe(true);
});

// ✅ Verifies rollback between tests
it('does not see the user created in the previous test', async () => {
  const GET_USERS = gql`
    query {
      users {
        nodes {
          id
          username
        }
      }
    }
  `;

  const fetchRes: any = await query(GET_USERS);

  expect(snapshot(fetchRes)).toMatchSnapshot('usersAfterRollback');
  expect(fetchRes.data.users.nodes).toHaveLength(0);
});

// ✅ Verifies context is set correctly
it('returns pg context settings from current_setting() function', async () => {
  db.setContext({ role: 'authenticated', 'myapp.user_id': '123' });
  
  const GET_CONTEXT = gql`
    query {
      currentRole: currentSetting(name: "role")
      userId: currentSetting(name: "myapp.user_id")
    }
  `;

  const res: any = await query(GET_CONTEXT);

  expect(snapshot(res)).toMatchSnapshot('pgContext');
  expect(res.data.currentRole).toBe('authenticated');
  expect(res.data.userId).toBe('123');
});

// ❌ Simulates access denied due to anonymous role
it('fails to access context-protected data as anonymous', async () => {
  const GET_CONTEXT = gql`
    query {
      currentRole: currentSetting(name: "role")
      userId: currentSetting(name: "myapp.user_id")
    }
  `;

  db.setContext({ role: 'anonymous' });

  const res: any = await query(GET_CONTEXT);

  expect(snapshot(res)).toMatchSnapshot('unauthorizedContext');
  expect(res.errors).toBeDefined();
  expect(res.errors[0]?.message).toMatch(/permission denied/i);
});
