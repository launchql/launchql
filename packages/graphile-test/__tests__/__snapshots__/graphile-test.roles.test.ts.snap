// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a user and fetches it: createUser 1`] = `
{
  "data": {
    "createUser": {
      "user": {
        "id": "[ID]",
        "username": "testuser1",
      },
    },
  },
}
`;

exports[`does not see the user created in the previous test: usersAfterRollback 1`] = `
{
  "data": {
    "users": {
      "nodes": [],
    },
  },
}
`;

exports[`fails to access context-protected data as anonymous: unauthorizedContext 1`] = `
{
  "data": {
    "currentRole": null,
    "userId": null,
  },
  "errors": [
    {
      "locations": [
        {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "permission denied for function current_setting",
      "path": [
        "currentRole",
      ],
    },
    {
      "locations": [
        {
          "column": 3,
          "line": 3,
        },
      ],
      "message": "current transaction is aborted, commands ignored until end of transaction block",
      "path": [
        "userId",
      ],
    },
  ],
}
`;

exports[`returns pg context settings from current_setting() function: pgContext 1`] = `
{
  "data": {
    "currentRole": "authenticated",
    "userId": "123",
  },
}
`;
